### PR TITLE
[ISSUE-21] Gibbs algorithm implementation

### PR DESCRIPTION
**Pull request by safronovD**
Implementation according to https://ocw.mit.edu/courses/biology/7-91j-foundations-of-computational-and-systems-biology-spring-2014/lecture-slides/MIT7_91JS14_Lecture9.pdf

- PWCalculator stores strategy
- PWM is built with MotifSet (not Fastafile)
- MotifSet stores possible motifs from Fastafile
- delta - square-law error

### CODE REVIEW

**Review by darina-ovis — APPROVED** [review]

File: src/main/java/ru/spbstu/calculator/MotifSet.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                .get();
+        data.set(data.indexOf(current), motif);
+    }
+
+    public enum Policy {
+        FIRST,
+        RANDOM
+    }
+
+    private void fillWithFirst(List<Motif> data) {
```
[the remark this answers is not in the document]
Comment (safronovD) [reply src/main/java/ru/spbstu/calculator/MotifSet.java]: done

File: src/main/java/ru/spbstu/calculator/PWCalculator.java [hunk 2]
```diff
@@ -1,45 +1,48 @@
...
     /**
      * Calculate position weight matrix with necessary output format
-     * @param fastaFile - fasta file with dataset
-     * @param strategy - strategy for calculation
+     * @param ms - MotifSet
      * @return calculated position weight matrix
      */
 
-    public PWMatrix calculateMatrix(FastaFile fastaFile, CalculationStrategy strategy) {
+    public PWMatrix calculateMatrix(MotifSet ms) {
```
[the remark this answers is not in the document]
Comment (safronovD) [reply src/main/java/ru/spbstu/calculator/PWCalculator.java]: done

File: src/main/java/ru/spbstu/calculator/PWMatrix.java [hunk 3]
```diff
@@ -27,23 +26,61 @@ public PWMatrix(int sequenceLength) {
...
+                s += '\t';
+            }
+            s += '\n';
+        }
+        return s;
+    }
+
+    public double countDelta(PWMatrix other) throws InvalidAlgorithmParameterException {
+        if (sequenceLength != other.getSequenceLength()) {
+            throw new InvalidAlgorithmParameterException();
```
[the remark this answers is not in the document]
Comment (safronovD) [reply src/main/java/ru/spbstu/calculator/PWMatrix.java:56]: moved this method to `GibbsMotifRecognizer`

File: src/main/java/ru/spbstu/calculator/ScoredMotif.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
+package ru.spbstu.calculator;
+
+public class ScoredMotif extends Motif implements Comparable<ScoredMotif> {
+    private static final int EPS_SCORE_COMPARE = 100;
+    // Position of the first symbol in parent record
```
[the remark this answers is not in the document]
Comment (safronovD) [reply src/main/java/ru/spbstu/calculator/ScoredMotif.java]: fixed

File: src/main/java/ru/spbstu/calculator/ScoredMotif.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        this.score = score;
+    }
+
+    public String getRecordID() {
+        return super.getRecordID();
+    }
+
+    public int getPosition() {
+        return super.getPosition();
+    }
```
[the remark this answers is not in the document]
Comment (safronovD) [reply src/main/java/ru/spbstu/calculator/ScoredMotif.java]: fixed

File: src/main/java/ru/spbstu/recognizer/GibbsMotifRecognizer.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            ScoredMotif maxScoredMotif = resultScore.getMax();
+
+            // Change found Motif with one with same recordID
+            motifSet.setMotif(maxScoredMotif);
+
+            // update delta
+            PWMatrix newMatrix = pwCalculator.calculateMatrix(motifSet);
+            delta = matrix.countDelta(newMatrix);
+            matrix = newMatrix;
+        }
```
[the remark this answers is not in the document]
Comment (safronovD) [reply src/main/java/ru/spbstu/recognizer/GibbsMotifRecognizer.java]: done

File: src/main/java/ru/spbstu/calculator/MotifSet.java [hunk 7]
```diff
@@ -9,39 +9,35 @@
 
 public class MotifSet {
 
-    private List<Motif> data;
-    private final int sequenceLength;
-    private final FastaFile fastaFile;
+    List<Motif> data;
+    final int sequenceLength;
+    final FastaFile fastaFile;
```
Comment (darina-ovis) [inline src/main/java/ru/spbstu/calculator/MotifSet.java:14]: why not private?

File: src/main/java/ru/spbstu/calculator/MotifSet.java [hunk 8]
```diff
@@ -9,39 +9,35 @@
...
         switch (policy) {
             case FIRST:
-                fillWithFirst(result);
+                fillWithFirst();
                 break;
             case RANDOM:
-                fillWithRandom(result);
+                fillWithRandom();
                 break;
         }
```
Comment (darina-ovis) [inline src/main/java/ru/spbstu/calculator/MotifSet.java]: What if policy is neither FIRST, nor RANDOM?

File: src/main/java/ru/spbstu/recognizer/GibbsMotifRecognizer.java [hunk 9]
```diff
@@ -11,49 +11,102 @@
...
+        private int prevInt = -1;
+        private Random random = new Random();
+
+        public int nextInt(int max) {
+            int nextRand = prevInt;
+
+            // to not repeat random number
+            while (nextRand == prevInt) {
+                nextRand = random.nextInt(max);
+            }
```
Comment (darina-ovis) [inline src/main/java/ru/spbstu/recognizer/GibbsMotifRecognizer.java:106]: prevInt contains only the previous value. So the random values can repeat. In order to have unique values we need to keep all used values but I don't think we really need it.

**Review by darina-ovis — APPROVED** [review]

File: src/main/java/ru/spbstu/calculator/MotifSet.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+    public void setMotif(Motif motif) {
+        Motif current = data.stream().
+                filter(m -> m.getRecordID().equals(motif.getRecordID())).
+                findAny()
+                .get();
+        data.set(data.indexOf(current), motif);
+    }
+
+    public enum Policy {
```
Comment (iporsh) [inline src/main/java/ru/spbstu/calculator/MotifSet.java:51]: Could you separate Policy from MotifSet

**Pull request merged**